### PR TITLE
Feat: add exclude files option

### LINKED ISSUE #58: Exclude specified paths from mutants generation
I believe that it would be handy to exclude specified paths from the mutants generation.

Let's say that I want to exclude things like configuration logic or some test helper logic files, located within the `src` directory. Right now the default behaviour of `cargo mutants` command is that it tries to generate mutants for every file in the `src`  directory and in their child directories.

I could envision having a command line parameter as follows:

```
cargo mutants --exclude <regexes-for-paths-to-be-excluded>
```

### Prerequisite:
Git clone [test repo](https://github.com/dmoka/test-repo-for-cargo-mutants.git)

### Steps:
1. Go to the main directory of the cloned repo
2. Execute the following command
```
cargo mutants --exclude "(\/src\/tests)" "(\/src\/config)"
```

### Expected behaviour:
Every file which is located in either `/src/tests` or `/src/config` directories are excluded from mutants generation.

### Actual behaviour:
Because of the default behaviour of `cargo-mutants` (and due to the non-existency of this command line parameter) there are mutants generated for the files both in `/src/tests` and `/src/config`

### PR DESCRIPTION
**Pull request by dmoka**
## Description
Added new CLI optional parameter `--exclude` to exclude files and paths, according to [issue](https://github.com/sourcefrog/cargo-mutants/issues/58).

Inspired by the `--file` parameter.

Fixes #58 

## Note
I was not able to debug any of the CLI tests, I wonder if it is normal or if something is off on my machine. 
- Because if it is normal, then maybe we could come up with a better debuggable testing strategy, of course if it is possible in an ergonomic way. It felt quite cumbersome to place print statements everywhere to debug my code when executing CLI tests
- If something needs to be configured to make those tests debuggable, then I feel that we should document it

Let me know what you think about this.

Anyway, it was a great fun to work with the codebase, especially liked this snapshot testing!

### CODE REVIEW

File: tests/snapshots/cli__list_mutants_well_tested_examine_name_filter.snap [hunk 1]
```diff
@@ -1,6 +1,5 @@
 ---
 source: tests/cli.rs
-assertion_line: 41
```
Comment (dmoka) [inline tests/snapshots/cli__list_mutants_well_tested_examine_name_filter.snap:3]: I wonder why this line has been removed. What I did is that I regenerated the snapshot in order to have new/refactored name for it.

**Comment by sourcefrog** [comment]
Thanks for the PR!

> I was not able to debug any of the CLI tests, I wonder if it is normal or if something is off on my machine.
> 
> * Because if it is normal, then maybe we could come up with a better debuggable testing strategy, of course if it is possible in an ergonomic way. It felt quite cumbersome to place print statements everywhere to debug my code when executing CLI tests
> * If something needs to be configured to make those tests debuggable, then I feel that we should document it

As you may have seen the CLI tests start a new invocation of `cargo mutants` in a subprocess. I think this is nice in that it does test the public interface in the most realistic way, and also gives good isolation between tests. But it does mean that if you set a breakpoint in the parent process it won't be hit in the child. 

The immediately concrete suggestion would be to just run the debugger on the command that you want to run inside the test, like `cargo mutants --exclude thing -d testdata/tree/something`. Once you have that working well, copy the command to a test. I realize that may not help if there is a bug that only reproduces when run inside the test, but again the approach of only running shell commands should hopefully make that rare?

If that doesn't work let me know?

Conceivably you could also arrange to hook in on every new invocation even in subprocesses but I haven't done that yet.

> Let me know what you think about this.
> 
> Anyway, it was a great fun to work with the codebase, especially liked this snapshot testing!

File: src/options.rs [hunk 2]
```diff
@@ -108,3 +101,21 @@ impl TryFrom<&Args> for Options {
         })
     }
 }
+
+fn build_glob_set(glob_set: &Vec<String>) -> Result<Option<GlobSet>>{
```
Comment (sourcefrog) [inline src/options.rs:105]: looks like this needs `cargo fmt` (or CI will fail)

Comment (sourcefrog) [inline src/options.rs:105]: even though it is fairly self-explanatory maybe add a comment?

Comment (sourcefrog) [inline src/options.rs:105]: I think this would be a reasonable thing to add unit tests for, in a `mod test` in this file: e.g. that files with and without slashes do and don't match in various combinations.

File: src/options.rs [hunk 3]
```diff
@@ -108,3 +101,21 @@ impl TryFrom<&Args> for Options {
...
+    let globset = if glob_set.is_empty() {
```
Comment (sourcefrog) [inline src/options.rs:106]: In Rust it would you don't need to assign to a variable just to return it; just let the `if` be the single expression in this function. (Clippy might say this?)

File: src/source.rs [hunk 4]
```diff
@@ -164,7 +165,8 @@ fn indirect_sources(
                     .expect("strip prefix")
                     .to_owned()
             })
-            .filter(|rel_path| globset.as_ref().map_or(true, |gs| gs.is_match(rel_path)))
+            .filter(|rel_path| examine_globset.as_ref().map_or(true, |gs| gs.is_match(rel_path)))
+            .filter(|rel_path| exclude_globset.as_ref().map_or(true, |gs| !gs.is_match(rel_path)))
```
Comment (sourcefrog) [inline src/source.rs:169]: We should document in the readme or option docstrings how these two interact. Matching inclusions before exclusions makes sense to me.

File: tests/cli.rs [hunk 5]
```diff
@@ -273,12 +273,21 @@ fn list_mutants_well_tested() {
 }
 
 #[test]
-fn list_mutants_well_tested_name_filter() {
+fn list_mutants_well_tested_examine_name_filter() {
```
Comment (sourcefrog) [inline tests/cli.rs:276]: probably we should have a test for using them both together, too?

File: tests/snapshots/cli__list_mutants_well_tested_examine_name_filter.snap (hunk 1 above)
Comment (sourcefrog) [reply tests/snapshots/cli__list_mutants_well_tested_examine_name_filter.snap:3]: I don't know!

Comment (sourcefrog) [reply tests/snapshots/cli__list_mutants_well_tested_examine_name_filter.snap:3]: It's probably harmless though.

File: tests/cli.rs [hunk 6]
```diff
@@ -273,12 +273,21 @@ fn list_mutants_well_tested() {
...
     run()
         .arg("mutants")
         .args(["--list", "--file", "nested_function.rs"])
         .current_dir("testdata/tree/well_tested")
-        .assert_insta("list_mutants_well_tested_name_filter");
+        .assert_insta("list_mutants_well_tested_examine_name_filter");
+}
+
+#[test]
+fn list_mutants_well_tested_exclude_name_filter() {
```
Comment (sourcefrog) [inline tests/cli.rs:285]: Probably we should also test, for at least one type of filter, how slashes are handled. I think `well_tested` has some submodules that you could use.

File: src/main.rs [hunk 7]
```diff
@@ -78,6 +78,11 @@ struct Args {
     #[argh(option, short = 'f')]
     file: Vec<String>,
 
+    /// glob for files to exclude; with no glob, all files are included; globs containing
```
Comment (sourcefrog) [inline src/main.rs:81]: We should also mention this in `README.md` and `NEWS.md`.

**Review by sourcefrog — CHANGES_REQUESTED** [review]
Thanks, looks good!

File: src/options.rs (hunk 2 above)
Comment (sourcefrog) [reply src/options.rs:105]: Actually CI will not yet reject it because it's not hooked up #67

File: src/options.rs [hunk 8]
```diff
@@ -108,3 +101,21 @@ impl TryFrom<&Args> for Options {
...
+        Ok(None)
+    } else {
+        let mut builder = GlobSetBuilder::new();
+        for glob_str in glob_set {
+            if glob_str.contains('/') {
```
Comment (sourcefrog) [inline src/options.rs:111]: Maybe this should also match backslash on Windows, by also checking against https://doc.rust-lang.org/std/path/constant.MAIN_SEPARATOR.html as well as `/`. We could then have a unit test in this file that runs on Windows only and that checks that it matches only paths.

File: tests/snapshots/cli__list_mutants_well_tested_examine_name_filter.snap (hunk 1 above)
Comment (dmoka) [reply tests/snapshots/cli__list_mutants_well_tested_examine_name_filter.snap:3]: all right, let's leave it then, hopefully, it won't come back.

**Comment by dmoka** [comment]
> Thanks for the PR!
> 
> > I was not able to debug any of the CLI tests, I wonder if it is normal or if something is off on my machine.
> > 
> > * Because if it is normal, then maybe we could come up with a better debuggable testing strategy, of course if it is possible in an ergonomic way. It felt quite cumbersome to place print statements everywhere to debug my code when executing CLI tests
> > * If something needs to be configured to make those tests debuggable, then I feel that we should document it
> 
> As you may have seen the CLI tests start a new invocation of `cargo mutants` in a subprocess. I think this is nice in that it does test the public interface in the most realistic way, and also gives good isolation between tests. But it does mean that if you set a breakpoint in the parent process it won't be hit in the child.
> 
> The immediately concrete suggestion would be to just run the debugger on the command that you want to run inside the test, like `cargo mutants --exclude thing -d testdata/tree/something`. Once you have that working well, copy the command to a test. I realize that may not help if there is a bug that only reproduces when run inside the test, but again the approach of only running shell commands should hopefully make that rare?
> 
> If that doesn't work let me know?
> 
> Conceivably you could also arrange to hook in on every new invocation even in subprocesses but I haven't done that yet.
> 
> We could add something about this to `CONTRIBUTING.md`.
> 
> > Let me know what you think about this.
> > Anyway, it was a great fun to work with the codebase, especially liked this snapshot testing!

That's indeed a great one, I also prefer targeting the public API in the tests, we can namely test use cases from the user point of view.

The issue with attaching a debugger to `cargo mutants --exclude thing -d testdata/tree/something` that the `exclude` param still does not exist for the `cargo mutants` command, therefore it fails. What I did is that I created a new Run configuration with CLion where I execute this command, but it resulted in exit code 1. I also tried with just `cargo mutants` but still did not stop at the breakpoint in the `main.rs`, although the mutants generation runs. It is possible that it is a limitation of CLion. What tool do you use for debugging separate commands, VSCode?


Anyway, improving the debuggability of the project would be out of scope of this PR, so we could indeed add it to CONTRIBUTING.md. I can add it there, if it is okay by you.

File: src/options.rs (hunk 2 above)
Comment (dmoka) [reply src/options.rs:105]: yes, this was my bubu, I will execute fmt (and also clippy) once everything else is processed :)

Comment (dmoka) [reply src/options.rs:105]: Maybe instead of adding a comment, we could also reveal more intent in the function name. 

How about this name: `build_glob_set_from_cli_arg_values` ?

Comment (dmoka) [reply src/options.rs:105]: Generally, I try to avoid writing tests coupled to the code structure and internal functions. The reason is that such things can result in fragile tests, so they can easily break in the future, in case of refactoring of the code.

What do you think about covering these cases in `cli.rs`? These are sort of edge cases, so I believe having public API tests for these can be reasonable.

File: src/options.rs (hunk 3 above)
Comment (dmoka) [reply src/options.rs]: good point, done!

File: tests/cli.rs (hunk 5 above)
Comment (dmoka) [reply tests/cli.rs:276]: excellent idea, I like it!

Comment (dmoka) [reply tests/cli.rs:276]: I think I will add a new testree for this with having child directories.

Because the default behaviour of `cargo mutants` is to check the `src` directory, but I want a case where I `examine` a child directory but `exclude` a specific file.

```
cargo mutants --file src/module/utils/*.rs --exclude nested_function.rs
```

File: tests/cli.rs (hunk 6 above)
Comment (dmoka) [reply tests/cli.rs:285]: Just to be sure: you mean testing for example excluding a folder like `cargo mutants --exclude /folder/*`, right?

Or what do you mean by `at least one type of filter`?

File: src/options.rs (hunk 8 above)
Comment (dmoka) [reply src/options.rs]: So basically, we just need to check only against [std](https://doc.rust-lang.org/std/index.html)::[path](https://doc.rust-lang.org/std/path/index.html)::[MAIN_SEPARATOR](https://doc.rust-lang.org/std/path/constant.MAIN_SEPARATOR.html#), which will be then determined based on the execution platform, right?

**Comment by sourcefrog** [comment]
Sorry `cargo mutants --exclude thing -d testdata/tree/something` was a bit misleading. That runs `cargo` which then runs in another subprocess whichever `cargo-mutants` binary it finds on your path. The tests arrange for it to find the right one, but if you run it from the shell or your debugger it will likely find the installed copy, not your new build. Also, the intermediate `cargo` process will get in the way of attaching the debugger.

What you probably actually want to debug is `target/debug/cargo-mutants mutants --exclude ...`, after building it.

Note you must have `mutants` as `argv[1]` because that's the protocol for cargo subcommands. 

This will run your new code as the direct child process and so the debugger should work.

I have used the VS Code debugger although I probably don't reach for it as often as would be optimal because I find the UI of configuring targets in json a bit weird. But this should work with anything that can launch and debug a child.

File: src/options.rs (hunk 8 above)
Comment (sourcefrog) [reply src/options.rs]: On Windows I think we should accept either. So 

`if glob_str.contains('/') || glob_str.contains(std::path::MAIN_SEPARATOR)`

File: tests/cli.rs (hunk 6 above)
Comment (sourcefrog) [reply tests/cli.rs:285]: > Just to be sure: you mean testing for example excluding a folder like cargo mutants --exclude /folder/*, right?

Right.

> Or what do you mean by at least one type of filter?

Either includes or excludes.

Like e.g. include one submodule, but then exclude one file within it.

File: src/options.rs (hunk 2 above)
Comment (sourcefrog) [reply src/options.rs:105]: Yep, I generally feel the same. 

In this case the public interface tests are a _little_ slow because they spawn subprocesses which can take a fraction of a second. If they just do `--list` that's not so bad, and that's probably enough to test this well.

Comment (sourcefrog) [reply src/options.rs:105]: maybe it's fine as it is

**Comment by dmoka** [comment]
> Sorry `cargo mutants --exclude thing -d testdata/tree/something` was a bit misleading. That runs `cargo` which then runs in another subprocess whichever `cargo-mutants` binary it finds on your path. The tests arrange for it to find the right one, but if you run it from the shell or your debugger it will likely find the installed copy, not your new build. Also, the intermediate `cargo` process will get in the way of attaching the debugger.
> 
> What you probably actually want to debug is `target/debug/cargo-mutants mutants --exclude ...`, after building it.
> 
> Note you must have `mutants` as `argv[1]` because that's the protocol for cargo subcommands.
> 
> This will run your new code as the direct child process and so the debugger should work.
> 
> I have used the VS Code debugger although I probably don't reach for it as often as would be optimal because I find the UI of configuring targets in json a bit weird. But this should work with anything that can launch and debug a child.

Awesome, I could successfully make it work with CLion, now I am able to debug the latest built code! 🙌 

Thanks a lot for the detailed explanation!

File: testdata/tree/with_child_directories/src/methods.rs [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+    }
+}
+
+impl fmt::Debug for &Foo {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "&Foo {}", self.i)
+    }
+}
+
+impl Default for Foo {
```
Comment (sourcefrog) [inline testdata/tree/with_child_directories/src/methods.rs:33]: Each function here will generate a mutation (or possibly more than one) and each of them will make the test suite slower if we actually test the mutations, especially on Windows. It won't be terrible but we might as well keep it relatively fast unless there's a reason.

I think this is copied from another tree that was trying to test mutation of impls, so it has several of them. 

If the point of this tree is to test directory structures then it's probably enough to just have one function per module.

Also it'll make the expected output a bit smaller and easier to read.

**Review by sourcefrog — CHANGES_REQUESTED** [review]
Looking good.

**Comment by sourcefrog** [comment]
Thanks! I guess our comments raced, but to be clear: I think the only remaining thing is to shrink the newly added testdata tree.

File: tests/cli.rs (hunk 6 above)
Comment (sourcefrog) [reply tests/cli.rs:285]: I see this is done, thanks!

**Comment by sourcefrog** [comment]
Actually, I'll merge this and then shrink the tests a bit. Thanks for the PR!

I agree the gh comments UX is a bit clunky.

File: testdata/tree/with_child_directories/src/methods.rs (hunk 9 above)
Comment (sourcefrog) [reply testdata/tree/with_child_directories/src/methods.rs:33]: I'll merge this and fix it separately.

**Review by sourcefrog — APPROVED** [review]

**Comment by dmoka** [comment]
> Actually, I'll merge this and then shrink the tests a bit. Thanks for the PR!
> 
> I agree the gh comments UX is a bit clunky.

Sorry, I was abroad last week with limited screen time, so I was not able to pick up your last improvement point.

Great, it was my pleasure to contribute! Thanks for the support! :pray:
